src/components/users/controllers.tsx

```diff
@@ -89,3 +96,119 @@ export async function getUser(
     ),
   };
 }
+
+export async function resetPasswordRequestToken(ctx: IContext, _params: IParameters): Promise<IResponse> {
+  const template = new Template(ctx.viewContext, 'Request password reset');
+
+  return await Promise.resolve({
+    body: template.render(<PasswordResetRequest csrf={ctx.viewContext.csrf} />),
+  });
+}
+
+export async function resetPasswordObtainToken(
+  ctx: IContext, params: IParameters, body: IPasswordResetBody,
+): Promise<IResponse> {
+  const VALID_EMAIL = /[^.]@[^.]/;
+  const email = body.email;
+  const template = new Template(ctx.viewContext, 'Request password reset');
+
+  if (!email || !VALID_EMAIL.test(email)) {
+    template.title = 'Error: Request password reset';
+
+    return {
+      body: template.render(<PasswordResetRequest csrf={ctx.viewContext.csrf} invalidEmail={true} />),
+      status: 400,
+    };
+  }
+
+  const uaa = new UAAClient({
+    apiEndpoint: ctx.app.uaaAPI,
+    clientCredentials: {
+      clientID: ctx.app.oauthClientID,
+      clientSecret: ctx.app.oauthClientSecret,
+    },
+  });
+
+  template.title = 'Password reset successfully requested';
+  const successPage = <PasswordResetSuccess
+    title="Password reset successfully requested"
+    message="You should receive an email shortly with further instructions."
+    footnote="If you do not receive an email in the next 24 hours, your account may not exist in the system or you could have setup authentication with Google or Microsoft Single Sign-on."
+  />;
+
+  const user = await uaa.findUser(email);
```

Comment:
Would it be worth making a call to the accounts API instead of UAA and explicitly returning a response:

```
const accountsUser = accountsClient.getUserByEmail(email)
if (!user) return "not found"

const uaaUser = uaaClient.getUserByGUID(user.guid)
if (uaaUser.origin != 'uaa') return "use single sign on"

const code = await uaa.obtainPasswordResetCode(uaaUser.userName);
...
```

My understanding is

- If the user is not using UAA to manage their account, UAA may not know their email
- PaaS accounts is the canonical store of email addresses
- If they are using SSO we should just tell them they are using SSO (as per the "It's done when" on the story)
- If we are worried about security we can introduce an artificial delay on this page to reduce the rate at which someone can enumerate email addresses.`

Reply:
Using the Accounts API is a sensible apporach here. Just wondering if there are any gotchas that we've not throught of that Rafal already has.

Reply:
The "It's done when" is:

> a SSO enabled tenant receives a helpful prompt on the page in response to asking for a password reset rather than nothing.

The current implementation of this PR is:

- try to find the user based on their email address in UAA
- if we cannot find the user or the user is using SSO, give them a page which says we have sent an email (which we haven't sent) and to wait 24h
- if we can find the user, and the user is not using SSO, send them an email

I think there are some improvements which block this from being merged:
- We should not say we have sent an email when we haven't sent an email
- If we cannot find the user we should say "your user was not found"
- If the user is using SSO we should say "you are using SSO"
- If we have security concerns we can artificially introduce a time delay in rendering the response, or we can send an email instead